Adds an alternate implementation for a jinja filter decorator

lorengordon opened the pull request:
### What does this PR do?

Adds a decorator, `@jinja_filter`, that will register the function as a jinja filter.

This PR represents an alternate approach to the decorator in PR #37808. In this PR, functions marked as jinja filters are added to a class attribute that maps the filter name to the function. The class attribute is then merged with the `jinja_env.filters` dictionary in `utils/templates.py`. As currently written, this approach works for functions only in the utils modules.

Getting this approach to work for execution modules involved looping over all the execution modules to force the lazy loader to apply the decorator, which may have negative performance impacts and so has been left out. However, for future reference in case someone wants to take on that project, in `salt/utils/template.py`, insert this just before updating the `jinja_env.filters` dictionary:

```
    for function_name in context['salt'].keys():
        # Simply iterating over the keys forces the execution modules to load
        # and process the jinja env decorator(s)
        pass
```

### What issues does this PR fix or reference?

Fixes #37791
Related #28236
Related #37769

### Tests written?

Partially. I've updated the existing tests to be compatible with these changes, but have not written new tests.

lorengordon commented on salt/utils/templates.py:
```diff
@@ -345,11 +346,17 @@ def render_jinja_tmpl(tmplstr, context, tmplpath=None):
     jinja_env.filters['yaml_squote'] = salt.utils.yamlencoding.yaml_squote
     jinja_env.filters['yaml_encode'] = salt.utils.yamlencoding.yaml_encode
 
+    for function_name in context['salt'].keys():
```
I'd like some guidance on this approach, and whether there is a better option to force the execution modules to load and process the decorators.

mirceaulinic replied:
How is this approach any different? I can see the same iteration over the ```__salt__``` object, which is the drawback of the other one.
An improvement would mean avoiding this IMO.

lorengordon replied:
Which is why I am asking if there is a better way to force the execution modules to load. That is the only reason this loop exists.

At the moment, the primary improvement is that this implementation works for `utils` functions.

mirceaulinic commented on salt/utils/network.py:
```diff
@@ -206,6 +207,7 @@ def is_ip(ip):
     return is_ipv4(ip) or is_ipv6(ip)
 
 
+@jinja_filter('is_ipv4')
```
As a side note, please do not override the existing ```is_ipv4``` filter: https://github.com/saltstack/salt/blob/develop/salt/utils/templates.py#L354 which is using the ```is_ipv4_filter``` functions.

lorengordon replied:
Sure, those are just there as working examples at the moment. Planning to remove those commits after the discussion. The only filters I'd include in this PR are those that already exist.

mirceaulinic commented on salt/utils/templates.py:
```diff
@@ -345,11 +346,17 @@ def render_jinja_tmpl(tmplstr, context, tmplpath=None):
     jinja_env.filters['yaml_encode'] = salt.utils.yamlencoding.yaml_encode
 
+    for function_name in context['salt'].keys():
+        # Simply iterating over the keys forces the execution modules to load
```
> Simply iterating over the keys forces the execution modules to load

I don't think so - when the template is rendered, the ```__salt__``` object is injected under the key ```salt``` of the context (e.g.: https://github.com/saltstack/salt/blob/7dd6d2a47fb755178c80a65a193d55517c78130e/salt/modules/file.py#L3614).
So the for loop is a simple iteration over the keys of a dictionary, whose keys are function objects.

mirceaulinic replied:
In the branch you linked two days ago: https://github.com/saltstack/salt/pull/37808#issuecomment-263161237 I saw the same piece of code, that yourself confirmed it did not work. Now it does? I will give it a whirl tomorrow.

lorengordon replied:
No, that loop did not exist previously. I added that this morning and cleaned up the commits as I had made a bit of a mess of the branch, trying to figure out why your implementation worked for execution modules and mine did not. Turns out the key factor was this loop.

lorengordon replied:
It works for execution modules due to that loop. It works for utils modules either way. I'm hoping someone from the salt team, more familiar with the loader, can tell me if it is possible to get the same result without the loop.

The benefit to this implementation is that we can use the same approach to adding jinja filters across the salt code base, rather than having one approach for execution modules and another approach for utils modules.

mirceaulinic replied:
Okay,  so now it works also for execution modules thanks to that iteration. I don't see any for the utils. But I can see that you are doing the same mistake as previously, rewriting the same object. I would be surprised if that would work with more than one function from the utils!

> The benefit to this implementation is that we can use the same approach to adding jinja filters across the salt code base, rather than having one approach for execution modules and another approach for utils modules.

I agree, that would be nice to have. At the same time, bear in mind that you create yet another object for each function having the decorator, as a new instance of your class. I think the performance of this is even more poor. Whilst the other one simply updates two properties of the existing objects.

lorengordon replied:
>Okay, so now it works also for execution modules thanks to that iteration. I don't see any for the utils.

No iteration loop is required for the utils modules because those functions are not dependent on the lazy loader. So, no, you won't see that.

>But I can see that you are doing the same mistake as previously, rewriting the same object. 

That is not a mistake. I did the reading you suggested on classes and decorators. It is not a new instance of the class. That is not how class attributes actually work. I did test this and can confirm that `JinjaFilter.salt_jinja_filters` is a dictionary that contains _all_ the decorated utils and execution module functions.

mirceaulinic replied:
Okay, this sounds good to me then! I will test your branch tomorrow and will let you know!

lorengordon replied:
I appreciate the constructive feedback and discussion! Thanks @mirceaulinic!

lorengordon commented:
It works for execution modules due to that loop. It works for utils modules either way. I'm hoping someone from the salt team, more familiar with the loader, can tell me if it is possible to get the same result without the loop.

The benefit to this implementation is that we can use the same approach to adding jinja filters across the salt code base, rather than having one approach for execution modules and another approach for utils modules.

lorengordon commented:
Here is a simple sls to test this PR:

```
# cat /srv/salt/test-37952.sls
{%- set my_string = "foo" %}
{%- set my_list = ["bar", ] %}
{%- set my_dict = {"baz": "qux"} %}
{%- set my_ip = "10.0.0.1" %}
{%- set my_digest = "foo" %}
{%- set my_tz = "America/Denver" %}

test sequence:
  cmd.run:
    - name: echo 'mystring={{ my_string|sequence|first }}, my_list={{ my_list|sequence|first }}, my_dict={{ my_dict|sequence|first }}'

test is_ipv4:
  cmd.run:
    - name: echo 'is_ipv4={{ my_ip|is_ipv4 }}'

test hashutil_digest:
  cmd.run:
    - name: echo 'my_digest={{ my_digest | hashutil_digest }}'

test timezone_compare:
  cmd.run:
    - name: echo 'my_tz={{ my_tz | timezone_compare }}'
```

damon-atkins commented:
Looks good, you will need @thatch45 input.

mirceaulinic commented:
Indeed it works for both execution modules functions and utils. But I'm still concerned - I think there would be too many objects created at each iteration:

And they are treated separately:

- utils:

```
[WARNING ] My name is strftime and I live at 0x7f614c2c3350
[WARNING ] Marking "strftime" as a jinja filter
[WARNING ] My name is is_ipv4 and I live at 0x7f614c0450d0
[WARNING ] My name is None and I live at 0x7f614734c410
[WARNING ] Marking "yaml_dquote" as a jinja filter
```

- execution modules:

```
2016-11-30 10:49:11,008 [salt.utils.decorators                                        ][WARNING ][30713] My name is hashutil_digest and I live at 0x7facdb38fc50
2016-11-30 10:49:11,008 [salt.utils.decorators                                        ][WARNING ][30713] Marking "hashutil_digest" as a jinja filter
2016-11-30 10:49:11,009 [salt.utils.decorators                                        ][WARNING ][30713] My name is timezone_compare and I live at 0x7facdb39fc50
2016-11-30 10:49:11,009 [salt.utils.decorators                                        ][WARNING ][30713] Marking "timezone_compare" as a jinja filter
```

Don't you think would be overkill creating hundreds of objects, whilst the sole advantage is collecting also functions from utils? To be honest, I don't see too many potential good candidates from utils. I would say it's a nice-to-have with poor performance & I think we should keep digging into this problem.

mirceaulinic commented:
To not be misunderstood: I am not at all against this - it is in the advantage for all of us to have a feature that performs good. But let's not rush and do it just for the sake of the change. So far this brings another useful feature that I like, but with the cost of decreasing performances (which anyway are not the best because of that iteration we all hate). I would propose to keep thinking, find a smart way and come with a _real_ improvement!

lorengordon commented:
> [WARNING ] My name is strftime and I live at 0x7f614c2c3350

@mirceaulinic can you share the code generating that log, so I know what object you are printing?

mirceaulinic commented:
```decorators/__init__.py```:

```python
class JinjaFilter(object):
    '''
    This decorator is used to specify that a function is to be loaded as a
    Jinja filter.
    '''
    salt_jinja_filters = {}

    def __init__(self, name=None):
        '''
        '''
        self.name = name
        log.warning('My name is {name} and I live at {addr}'.format(
            name=name,
            addr=hex(id(self))))

    def __call__(self, function):
        '''
        '''
        name = self.name or function.__name__
        if name not in self.salt_jinja_filters:
            log.warning('Marking "{0}" as a jinja filter'.format(name))
            self.salt_jinja_filters[name] = function
        return function
```

Here you go.

lorengordon commented:
Thanks. Well, we can compare execution times. If I checkout my branch at the first commit, https://github.com/saltstack/salt/pull/37952/commits/a28a6d8a8ad34d28af50ff4943ad3ff8b28124ba, the decorator exists but no functions are marked. If I execute a simple state exercising a few of the pre-existing jinja filters, and I get an execution time of:

```
Total states run:     3
Total run time:  13.098 ms
```

If I checkout the HEAD of my branch, where every util function and a couple execution modules have been wrapped using this decorator, and I run the same state, I get an execution time of:

```
Total states run:     3
Total run time:  12.772 ms
```

So, a little bit faster, but this is only one run each and so not a great analysis, but at least we're not orders of magnitude slower out of the gate. Obviously, perhaps as we add more and more functions as filters, this might change. Other relative details: I'm running this on a t2.micro EC2 instance in AWS, and the OS is CentOS 6.

Here is the test state:

```
# cat /srv/salt/test-37952.sls
{%- set my_string = "foo" %}
{%- set my_list = ["bar", ] %}
{%- set my_dict = {"baz": "qux"} %}
{%- set my_ip = "10.0.0.1" %}
{%- set my_digest = "foo" %}
{%- set my_tz = "America/Denver" %}
{%- set my_regex = "https://example.com?foo=bar%20baz" %}

test sequence:
  cmd.run:
    - name: echo 'mystring={{ my_string|sequence|first }}, my_list={{ my_list|sequence|first }}, my_dict={{ my_dict|sequence|first }}'

test is_ipv4:
  cmd.run:
    - name: echo 'is_ipv4={{ my_ip|is_ipv4 }}'

test escape_regex:
  cmd.run:
    - name: echo 'my_regex={{ my_regex | regex_escape }}'
```

mirceaulinic commented:
> we can compare execution times

Not relevant, it is not linear. How many states did you run concomitantly? One? With performance I mean thousands of states run every second. Dealing with few thousands of memory allocations and throwing them immediately to the garbage is not the most optimal thing we can do - we must be careful when playing with the memory.

lorengordon commented:
You are totally welcome to design a test you think will be more accurate.

mirceaulinic commented:
> You are totally welcome to design a test you think will be more accurate.

I thought _you_ are trying to prove this works better :)

lorengordon commented:
I figure we're _both_ trying to understand the performance impacts of each implementation. ;)

If you want to flesh out some of the test requirements and dimensions, I can see what I can do.

mirceaulinic commented:
> I figure we're both trying to understand the performance impacts of each implementation. ;)

Can't agree more - I've never tested a new Salt core feature before.

> If you want to flesh out some of the test requirements and dimensions, I can see what I can do.

I have a few thousands of devices available, but I won't use them for testing purposes. But I can make some deeper checks with a fair amount of them somewhere at the end of next week.

lorengordon commented on salt/utils/templates.py:
```diff
@@ -345,20 +346,12 @@ def render_jinja_tmpl(tmplstr, context, tmplpath=None):
+    for function_name in context['salt'].keys():
+        # Simply iterating over the keys forces the execution modules to load
+        # and process the jinja env decorator(s)
+        pass
```
I rebased on latest develop, so the prior comment on this loop is hidden. I'd still like to get feedback from someone on the salt team more familiar with the loader, regarding this loop and whether there is an easier/better way to get the execution modules to load and process the decorator.

cachedout replied:
This has the potential to cause problems. It will be very slow and this is already a performance-sensitive area of the code. I have some ideas about how to improve this though and I'll try to see about modifying the LazyLoader class this week.

cachedout commented:
We need to figure out what to do with this one. @lorengordon and @mirceaulinic what are your thoughts?

lorengordon commented:
I think the concept simplifies the addition of new jinja filters by collecting the functions in one place and looping over them, which also cleans up `utils/templates.py` quite a bit. It's basically the same thing that the `@depends` decorator does, afaict.

I removed the part of this that enabled salt module functions to be used as a jinja filter, which I think was the main point of concern around performance (as that required looping over `context[salt]` to load all the modules).

So, I'm pretty happy with it as is, especially with develop as the target branch.

cachedout commented:
@lorengordon You'd be comfortable with this going into Nitrogen that's going to be released next month then?

mirceaulinic commented:
It sounds good to me, as long as we don't traverse the whole `salt` object (or any other kind of iteration). I didn't perform the profiling unfortunately, but i don't see any potential donwside, however i would encourage further testing.

lorengordon commented:
Heh, I'm confident in the code's functionality, but I'm not really in a position to test the performance extensively, so if I were you, I'd probably wait until after cutting the branch for the Nitrogen release to give it more time to cure in develop.

cachedout commented:
@lorengordon Gotcha. We'll hold on this. Thanks for the quick response.

lorengordon committed:
Adds a decorator for registering jinja filters

The decorator adds the filter to a class attribute, `salt_jinja_filters`
that maps filter names to functions.

lorengordon committed:
Applies the decorator to all jinja filters

lorengordon committed:
Moves regex_escape and unique filters out of the SerializerExtension class

lorengordon committed:
Updates tests to get custom filters from decorator class

lorengordon committed:
Updates jinja template tests with salt context

lorengordon commented:
@cachedout with the nitrogen branch cut, this is probably good to merge to develop now?

cachedout commented:
@lorengordon We can do that, sure. We'll need the merge conflicts resolved first, though, please. :]

lorengordon commented:
Something else must have gotten merged since I pinged you... I had _just_ rebased so it would merge cleanly. I'll take another look tonight when I'm back at a computer.

lorengordon commented:
@cachedout rebased and fixed the merge conflict. Should be good once the tests finish.
